brain/detail/circuit.h

```diff
@@ -214,20 +183,384 @@ class Circuit::Impl
 
     virtual Vector3fs getPositions(const GIDSet& gids) const = 0;
     virtual size_ts getMTypes(const GIDSet& gids) const = 0;
-    virtual Strings getMorphologyNames() const = 0;
+    virtual Strings getMorphologyTypeNames() const = 0;
     virtual size_ts getETypes(const GIDSet& gids) const = 0;
     virtual Strings getElectrophysiologyNames() const = 0;
     virtual Quaternionfs getRotations(const GIDSet& gids) const = 0;
     virtual Strings getMorphologyNames(const GIDSet& gids) const = 0;
 
     URI getMorphologyURI(const std::string& name) const
     {
-        URI uri(_morphologySource);
+        if (boost::filesystem::path(name).is_absolute())
+            return URI(name);
+
+        URI uri(getMorphologySource());
         uri.setPath(uri.getPath() + "/" + name + ".h5");
         return uri;
     }
 
-    const brion::SynapseSummary& getSynapseSummary() const
+    virtual URI getMorphologySource() const = 0;
+
+    virtual const brion::URI& getCircuitSource() const = 0;
+    virtual void saveMorphologyToCache(
+        const std::string& uri, const std::string& hash,
+        neuron::MorphologyPtr morphology) const = 0;
+
+    virtual CachedMorphologies loadMorphologiesFromCache(
+        const std::set<std::string>& hashes) const = 0;
+    virtual void saveSynapsePositionsToCache(
+        const uint32_t gid, const std::string& hash,
+        const brion::SynapseMatrix& value) const = 0;
+    virtual CachedSynapses loadSynapsePositionsFromCache(
+        const Strings& keys) const = 0;
+    virtual void _findSynapsePositionsColumns() const = 0;
+    virtual const brion::SynapseSummary& getSynapseSummary() const = 0;
+    virtual const brion::Synapse& getSynapseAttributes(
+        const bool afferent) const = 0;
+    virtual const brion::Synapse& getAfferentProjectionAttributes(
+        const std::string& name) const = 0;
+    virtual const brion::Synapse* getSynapseExtra() const = 0;
+    virtual const brion::Synapse& getSynapsePositions(
+        const bool afferent) const = 0;
+    virtual const URI& getSynapseSource() const = 0;
```

Comment:
I think I'm going to do another some other refactorings here also. This design is looking terrible to me. This will also remove the need for the compilation hacks you made below.